File: main/src/main/java/com/sedmelluq/lavaplayer/core/source/youtube/YoutubeLinkRouter.java
package com.sedmelluq.lavaplayer.core.source.youtube;

public interface YoutubeLinkRouter {
  <T> T route(String link, Routes<T> routes);

  interface Routes<T> {
    T track(String videoId);

    T playlist(String playlistId, String selectedVideoId);

    T mix(String mixId, String selectedVideoId);

    T search(String query);

    T searchMusic(String query);

    T anonymous(String videoIds);

    T none();
  }
}

File: main/src/main/java/lavaplayer/source/youtube/YoutubeLinkRouter.kt
package lavaplayer.source.youtube

interface YoutubeLinkRouter {
    fun <T> route(link: String, routes: Routes<T>): T

    interface Routes<T> {
        fun track(videoId: String): T

        fun playlist(playlistId: String, selectedVideoId: String?): T

        fun mix(mixId: String, selectedVideoId: String?): T

        fun search(query: String): T

        fun searchMusic(query: String): T

        fun anonymous(videoIds: String): T

        fun none(): T
    }
}
